### DISCUSSION TITLE
Question about likelihood (mean score？)

### DISCUSSION THREAD

**Discussion by nova8761** [talmolab/sleap#1323] [Help!]
When I was using DLC before, a likelihood (confidence) was given to each frame after training or prediction, which was a value between 0 and 1, meaning the probability of being correct was between 0% and 100%.
Now I'm going to try to use SLEAP instead. I find that SLEAP has a similar parameter called "mean score" for training and "score" for inference, but it seems to be a value between 0 and 2.
May I ask: How is this value defined and how is it calculated? It does not seem to be a probability value?

**Comment by roomrys** [comment]
Hi @nova8761,

The mean score is the mean of all the instance scores in the frame. The calculation of SLEAP's instance scores differ from type of model to type of model (see `Predictor`s below). In general, the instance score aims to measure the grouping of nodes (predicted points) into an instance. For most predictors, SLEAP uses the likelihood as a score which is proportional to the probability of a point being correct, but not normalized. Thus the higher the score the more "confident" SLEAP is in that prediction/grouping, but this confidence is not constrained between 0 to 1. That being said, one downfall of having a nonnormalized score means that it is difficult to compare predictions across different types of models.

Thanks,
Liezl

---

Related #1268 (and #1219)

> Hi @druz, @mollywestbrook,
> 
> ## Point
> The score comes directly from the output of our keras model (heatmaps that indicate the likelihood of each pixel containing the center of a particular keypoint). Each node will get its own heatmap (aka confidence map) and within each heatmap there will be multiple peaks for where SLEAP thinks there is some non-zero likelihood that a pixel contains the center of a node given the annotated frames. Note that this likelihood is not normalized, but is proportional to the joint probability of a "pixel containing the center of a particular keypoint" given the annotated data. This proportionality (and comparison to other likelihoods) is all we need to determine which peak in our heatmap to use as the predicted point for that node. This is done for all nodes and the likelihood is then used as the score for that point.
> 
> ## Instances
> Per the doc strings, the instance score is supposed to measure the grouping of nodes.
> 
> https://github.com/talmolab/sleap/blob/b5aab0edfbcfddce82e8ee02d9bd383604b9de93/sleap/instance.py#L1166-L1167
> 
> ### `SingleInstancePredictor`
> The score is a sum of all the point scores (not a grouping score because there is not grouping to be done - all predicted nodes must belong to this single instance). This score would not even fall within the range of the other `Predictor`s since it would be scaled by the number of nodes.
> 
> ### `TopDownPredictor`
> Much like the score for a point, this score comes from the heatmap of the first model in top-down (the centroid model) which finds the centroid/"anchor point" (the location of the node closest to the centroid of the object being tracked). Thus the instance score is the likelihood of a pixel containing the center of the "anchor point".
> 
> ### `BottomUpPredictor`
> The instance score for a `BottomUpPredictor` is the sum of all the edge scores in an instance. The edge score comes from the PAFs and is explained in the doc strings for `score_paf_lines` as:
> 
> https://github.com/talmolab/sleap/blob/b5aab0edfbcfddce82e8ee02d9bd383604b9de93/sleap/nn/paf_grouping.py#L354-L361
> 
> ### `TopDownMulticlassPredictor`
> The multiclass predictors have tracking baked in as another network that uses annotated `Track`s to predict instance tracks. So, now the instance score is the classification probabilities that an instance is in the assigned `Track`. Similar to the point scores, these classification probabilities are directly output from the keras model used to run inference. Note that these are referred to as probabilities not likelihoods; thus we expect them to fall within 0 to 1.
> 
> ### `BottomUpMulticlassPredictor`
> The same as the `TopDownMulticlassPredictor`
> 
> ### `MoveNetPredictor`
> The same as `SingleInstancePredictor`.
> 
> ## Tracking
> So the tracking score is derived using both an instance similarity score to form a cost matrix and then using a matching method to choose which instances to assign to which track (essentially selecting an element in the cost matrix to assign to the instance as a tracking score). The similarity functions to choose from are either instance similarity, centroid distance, or instance Intersection Over Union (IOU). The output from each of these will form the cost matrix. The method chosen for matching will then select which element of the cost matrix to assign each to each `Instance.tracking_score`.
> 
> https://github.com/talmolab/sleap/blob/b5aab0edfbcfddce82e8ee02d9bd383604b9de93/sleap/nn/tracker/components.py#L29-L78
